[iOS] Image context menu presentation should only be gated on text recognition results

whsieh opened the pull request:
#### a8666b738a4541517d50a52afa73d5efc9f85e5b
<pre>
[iOS] Image context menu presentation should only be gated on text recognition results
<a href="https://bugs.webkit.org/show_bug.cgi?id=242490">https://bugs.webkit.org/show_bug.cgi?id=242490</a>
rdar://94488144

Reviewed by Tim Horton and Devin Rousso.

When long pressing images on iOS 15+, context menu presentation is gated behind image analysis
results for two reasons:

1.  If the user long presses text that appears inside of an image element, we perform text selection
    instead of showing a context menu.

2.  If the image doesn&apos;t contain text at the touch location, we still wait for visual search,
    machine-readable code detection and (new in iOS 16) object lifting to finish before showing the
    context menu, in order to determine whether or not we should include &quot;Look Up&quot;, the entire QR
    code submenu, as well as &quot;Copy Subject&quot;.

With the inclusion of &quot;Copy Subject&quot; on iOS 16, this means that it can take an unexpectedly long
time for the context menu to show up when long pressing images. To mitigate this problem, we adjust
our gating logic so that the context menu may appear immediately after step (1), if the user isn&apos;t
long pressing text inside of an image.

As the context menu is showing, we simultaneously run image analysis (visual look up, MRC and object
lifting); if analysis finishes before the context menu has finished presenting, we show the context
menu with all of the image analysis actions up front; otherwise, use `-updateVisibleMenuWithBlock:`
to dynamically insert the new image analysis context menu items, as long as the WebKit client hasn&apos;t
overridden default context menu items.

* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(elementActionTypeToUIActionIdentifier):
* Source/WebKit/UIProcess/API/Cocoa/_WKElementActionInternal.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForLinkSheet:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Pull out all image-analysis-related context menu data (e.g. whether to include &quot;Show Text&quot; or &quot;Look
Up&quot;) into a separate struct, rather than as separate members. This allows us to more easily
accumulate results as each of the analyses finish, and also makes it easier to entirely reset this
state when beginning or ending context menu presentation.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView hasSelectableTextForImageContextMenu]):
(-[WKContentView hasVisualSearchResultsForImageContextMenu]):
(-[WKContentView copySubjectResultForImageContextMenu]):
(-[WKContentView machineReadableCodeSubMenuForImageContextMenu]):

Add helper methods to consult whether or not each of the image-analysis-related context menu actions
should be available. Use these in place of checking the individual ivars when determining whether or
not we should show these menu items.

(-[WKContentView presentVisualSearchPreviewControllerForImage:imageURL:title:imageBounds:appearanceActions:]):
(-[WKContentView actionSheetAssistant:shouldIncludeShowTextActionForElement:]):
(-[WKContentView actionSheetAssistant:shouldIncludeLookUpImageActionForElement:]):
(-[WKContentView _setUpImageAnalysis]):
(-[WKContentView _tearDownImageAnalysis]):
(-[WKContentView updateImageAnalysisForContextMenuPresentation:]):
(-[WKContentView imageAnalysisGestureDidBegin:]):
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):

This is the bulk of the change; we now call `-_invokeAllActionsToPerformAfterPendingImageAnalysis:`
up front, and accumulate the results in a `ImageAnalysisContextMenuActionData`. When all of the
analysis types have completed, we then set `_imageAnalysisContextMenuActionData` and update the
context menu with the corresponding items if needed.

(-[WKContentView imageAnalysisGestureDidTimeOut:]):
(-[WKContentView actionSheetAssistantShouldIncludeCopySubjectAction:]):
(-[WKContentView actionSheetAssistant:copySubject:sourceMIMEType:]):
(-[WKContentView continueContextMenuInteraction:]):

Unset the `_canUpdateVisibleContextMenuWithImageAnalysisActions` flag, if the WebKit client has
specified its own context menu configuration. Later on, if image analysis completes and this flag is
`NO`, we&apos;ll skip updating the visible menu to contain image analysis actions.

(-[WKContentView contextMenuInteraction:willEndForConfiguration:animator:]):
(-[WKContentView _updateContextMenuForMachineReadableCodeForImageAnalysis:]): Deleted.

Canonical link: <a href="https://commits.webkit.org/252307@main">https://commits.webkit.org/252307@main</a>
</pre>

dcrousso commented on Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
```diff
@@ -551,15 +551,13 @@ - (void)_appendOpenActionsForURL:(NSURL *)url actions:(NSMutableArray *)defaultA
-        }
 #endif
 #if ENABLE(IMAGE_ANALYSIS)
-        if ([_delegate respondsToSelector:@selector(actionSheetAssistant:shouldIncludeShowTextActionForElement:)] && [_delegate actionSheetAssistant:self shouldIncludeShowTextActionForElement:elementInfo])
```
does this actually need to change?

whsieh replied:
It doesn't, but I wanted to clean this up while I was touching this code since the element is unused anyways. I can split this out into a followup if you prefer.

dcrousso commented on Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
```diff
@@ -551,15 +551,13 @@ - (void)_appendOpenActionsForURL:(NSURL *)url actions:(NSMutableArray *)defaultA
 #endif
 #if ENABLE(IMAGE_ANALYSIS)
-        if ([_delegate respondsToSelector:@selector(actionSheetAssistant:shouldIncludeShowTextActionForElement:)] && [_delegate actionSheetAssistant:self shouldIncludeShowTextActionForElement:elementInfo])
+        if ([_delegate respondsToSelector:@selector(actionSheetAssistantShouldIncludeShowTextAction:)] && [_delegate actionSheetAssistantShouldIncludeShowTextAction:self])
```
does these actually need to change?

whsieh replied:
(Ditto)

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10965,23 +10958,71 @@ - (void)imageAnalysisGestureDidBegin:(WKImageAnalysisGestureRecognizer *)gesture
+            auto *elementInfo = [_WKActivatedElementInfo activatedElementInfoWithInteractionInformationAtPosition:strongSelf->_positionInformation userInfo:nil];
+            auto addActionIfNeeded = [&](_WKElementActionType action) {
+                auto actionIdentifier = elementActionTypeToUIActionIdentifier(action);
+                if (!actionIdentifier || [existingActionIdentifiers containsObject:actionIdentifier])
```
Instead of looking through the existing actions to see if we already have one that matches, what about if we have some extra `BOOL` inside the `ImageAnalysisContextMenuActionData` that indicate whether or not we should add the relevant item to the contextmenu?

e.g. in addition to `hasSelectableTextForImageContextMenu `, we could have a `shouldAddImageExtractionContextMenuItem` (and maybe a `didAddImageExtractionContextMenuItem`) that is called anytime we're about to add a `_WKElementActionTypeImageExtraction` item

whsieh replied:
After discussing this over Slack, and agreed on a solution where we:
- Add something to tell if the context menu has been overridden via delegate
- if so, then don’t add these image analysis actions at all
- if not, then add them without trying to mitigate duplicate IDs

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12274,9 +12314,7 @@ - (void)contextMenuInteraction:(UIContextMenuInteraction *)interaction willEndFo
-#if ENABLE(IMAGE_ANALYSIS_FOR_MACHINE_READABLE_CODES)
-    _contextMenuForMachineReadableCode.clear();
-#endif // ENABLE(IMAGE_ANALYSIS_FOR_MACHINE_READABLE_CODES)
+    _imageAnalysisContextMenuActionData = std::nullopt;
```
This used to just clear the MRC menu, but will now also remove the "copy subject" image.  Is that OK?

whsieh replied:
That's an excellent catch — while I had confirmed locally that "Copy Subject" still works, I'll do some more investigation to confirm that we can rely on this always being cleared out after we've already invoked the action handler.

whsieh replied:
So given UIKit's current implementation, it's guaranteed that the action is going to be invoked before the `-willEndForConfiguration:`. However, it does seem slightly fragile to rely on this ordering — I think a more robust way to deal with this state reset is to actually lazily reset it when we're about to compute `_imageAnalysisContextMenuActionData`, inside `-_completeImageAnalysisRequestForContextMenu:` (so that way we never end up using stale results).

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10994,18 +11035,17 @@ - (void)_completeImageAnalysisRequestForContextMenu:(CGImageRef)image requestIde
-#endif
-#if USE(UICONTEXTMENU) && ENABLE(IMAGE_ANALYSIS_FOR_MACHINE_READABLE_CODES)
-        [strongSelf _updateContextMenuForMachineReadableCodeForImageAnalysis:result];
+        data->machineReadableCodeMenu = result.mrcMenu;
```
I think we may still want to guard this behind `ENABLE(IMAGE_ANALYSIS_FOR_MACHINE_READABLE_CODES)`

whsieh replied:
Fixed!

dcrousso commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -11049,41 +11089,41 @@ - (void)imageAnalysisGestureDidTimeOut:(WKImageAnalysisGestureRecognizer *)gestu
             if (!strongSelf)
                 return;
 
+
```
Style: double newline

whsieh replied:
Fixed, good catch!

whsieh committed:
[iOS] Image context menu presentation should only be gated on text recognition results
https://bugs.webkit.org/show_bug.cgi?id=242490
rdar://94488144

Reviewed by Tim Horton and Devin Rousso.

When long pressing images on iOS 15+, context menu presentation is gated behind image analysis
results for two reasons:

1.  If the user long presses text that appears inside of an image element, we perform text selection
    instead of showing a context menu.

2.  If the image doesn't contain text at the touch location, we still wait for visual search,
    machine-readable code detection and (new in iOS 16) object lifting to finish before showing the
    context menu, in order to determine whether or not we should include "Look Up", the entire QR
    code submenu, as well as "Copy Subject".

With the inclusion of "Copy Subject" on iOS 16, this means that it can take an unexpectedly long
time for the context menu to show up when long pressing images. To mitigate this problem, we adjust
our gating logic so that the context menu may appear immediately after step (1), if the user isn't
long pressing text inside of an image.

As the context menu is showing, we simultaneously run image analysis (visual look up, MRC and object
lifting); if analysis finishes before the context menu has finished presenting, we show the context
menu with all of the image analysis actions up front; otherwise, use `-updateVisibleMenuWithBlock:`
to dynamically insert the new image analysis context menu items, as long as the WebKit client hasn't
overridden default context menu items.

* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(elementActionTypeToUIActionIdentifier):
* Source/WebKit/UIProcess/API/Cocoa/_WKElementActionInternal.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForLinkSheet:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:

Pull out all image-analysis-related context menu data (e.g. whether to include "Show Text" or "Look
Up") into a separate struct, rather than as separate members. This allows us to more easily
accumulate results as each of the analyses finish, and also makes it easier to entirely reset this
state when beginning or ending context menu presentation.

* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView hasSelectableTextForImageContextMenu]):
(-[WKContentView hasVisualSearchResultsForImageContextMenu]):
(-[WKContentView copySubjectResultForImageContextMenu]):
(-[WKContentView machineReadableCodeSubMenuForImageContextMenu]):

Add helper methods to consult whether or not each of the image-analysis-related context menu actions
should be available. Use these in place of checking the individual ivars when determining whether or
not we should show these menu items.

(-[WKContentView presentVisualSearchPreviewControllerForImage:imageURL:title:imageBounds:appearanceActions:]):
(-[WKContentView actionSheetAssistant:shouldIncludeShowTextActionForElement:]):
(-[WKContentView actionSheetAssistant:shouldIncludeLookUpImageActionForElement:]):
(-[WKContentView _setUpImageAnalysis]):
(-[WKContentView _tearDownImageAnalysis]):
(-[WKContentView updateImageAnalysisForContextMenuPresentation:]):
(-[WKContentView imageAnalysisGestureDidBegin:]):
(-[WKContentView _completeImageAnalysisRequestForContextMenu:requestIdentifier:hasTextResults:]):

This is the bulk of the change; we now call `-_invokeAllActionsToPerformAfterPendingImageAnalysis:`
up front, and accumulate the results in a `ImageAnalysisContextMenuActionData`. When all of the
analysis types have completed, we then set `_imageAnalysisContextMenuActionData` and update the
context menu with the corresponding items if needed.

(-[WKContentView imageAnalysisGestureDidTimeOut:]):
(-[WKContentView actionSheetAssistantShouldIncludeCopySubjectAction:]):
(-[WKContentView actionSheetAssistant:copySubject:sourceMIMEType:]):
(-[WKContentView continueContextMenuInteraction:]):

Unset the `_canUpdateVisibleContextMenuWithImageAnalysisActions` flag, if the WebKit client has
specified its own context menu configuration. Later on, if image analysis completes and this flag is
`NO`, we'll skip updating the visible menu to contain image analysis actions.

(-[WKContentView contextMenuInteraction:willEndForConfiguration:animator:]):
(-[WKContentView _updateContextMenuForMachineReadableCodeForImageAnalysis:]): Deleted.

Canonical link: https://commits.webkit.org/252307@main
